### PR TITLE
delegate gradio events via a custom event dispatcher

### LINKED ISSUE #4841: Gradio starting with 3.36 is extremely slow
### Describe the bug

I just upgraded my app from Gradio 3.32.0 to 3.36.1 (I cannot use versions in-between due to breaking change that was fixed post 3.35.2) and browser page load time grew from 5sec to 15sec.

Yes, its a highly complex app with too many controls, but 300% performance degradation is not what I would expect.

Once page load completes, actual functionality is on-par and I don't see any errors.
Note that assets download is not an issue, they all complete fast (seen in browser inspector network tab), its actual building of the UI, for example JS code  inside `http://127.0.0.1:7860/assets/Blocks-7b1f1558.js` takes 5sec to execute alone.

### Have you searched existing issues?  🔎

- [X] I have searched and found no existing issues

### Reproduction

1. Install [SD.Next](https://github.com/vladmandic/automatic) and start it using `webui.sh/webui.bat`  
  (first start will install a lot of dependencies so it may take long)
2. Change `requirements.txt` to use new Gradio and start again

### Screenshot

_No response_

### Logs

```shell
No error messages in console or browser console
```

### System Info

```shell
Server Gradio 3.36.1 on Ubuntu 22.04
Client Chrome 114 on Windows 11
```

### Severity

Blocking usage of gradio

### PR DESCRIPTION
**Pull request by pngwn**
## Description

This PR includes a few performance improvements.

- We know longer attach events manually for component events or share error events. Instead we delegate to a single listener at the very top of the DOM tree. In order to do this we need events to bubble (which svelte events don't), so I have introduced a `gradio.dispatch` method that gets passed to all components. Component can trigger events. just as they did before.
-  `gr.Markdown` and Markdown inside of dataframes was rerendering infinitely, so i fixed that.

https://github.com/gradio-app/gradio/assets/12937446/1989b06a-26ba-43b0-8663-6650cff590e5

- the model3d was rerendering several times per update which was both annoying and also quite expensive. I fixed that.
- handle mount was doing a lot of duplicate work (because it got called every time a component mounted), this was very problematic in large applications (like SD.Next + SDWebUI). I refactored some stuff so we only call `handle_mount` once when everything is done. This gives us about a 30% speed up.
- I also created a `Gradio` class for components, an instance of which is passed down to each components. This contains some app state (theme_mode) as well as two utilities (`dispatch` for triggering events, and `i18n`). Only dispatch is used right now.

All in all this reverts the regressions that were recently introduced into gradio + gives us a reasonable performance boost. I did some basic profile and For reasonably sized applications, things are probably about 30% faster (plus the regression fix). SD.Next now loads of me in about ~2 seconds (without dev tools open), on 3.32.0 it takes ~3. On 3.40.1 it takes about 8/9.

I haven't used the chrome profiler for a while but we are now best friends again.

There are more performance improvements we can make but they are either specific to individual components or some variation on 'stop rendering everything' (which is very much the plan but not something we can do right now).

Closes #4841.

### CODE REVIEW

File: demo/hello_blocks/run.py
[the remark this answers is not in the document]
Comment (pngwn) [reply demo/hello_blocks/run.py:1]: Yeah, sorry I meant to change that back. Was just testing.

**Comment by pngwn** [comment]
So I did a little testing with [SD.Next](https://github.com/vladmandic/automatic).

- The current version uses gradio 3.32.0 (the last 'working' version) and the load time is around 7-8 seconds.
- Main or latest take about 15-16 seconds to load.
- I installed this PR into SD.Next and the loading time seems to be back down to 7-8 seconds.

This isn't fast for these large applications, but it is never going to be. It does, however, seem to address the performance regression that was introduced in 3.36.0.

It should also fix the issue when switching between static + interactive modes.

There is additional work we need to do in order to really get these complex apps in a good place but we can take a closer look at that later. This PR certainly makes doing that a _lot_ easier because one of the big challenges was always adding and removing events as the UI changed.

**Comment by abidlabs** [comment]
@pngwn I can confirm that the interactivity issue is fixed.

However, I'm not able to reproduce the improvements in performance. In fact, there might be a degradation in performance.

Compare the loading times of the Spaces preview for example:
* This branch: https://huggingface.co/spaces/gradio-pr-deploys/pr-5279-all-demos
* Similar to main: https://huggingface.co/spaces/gradio-pr-deploys/pr-5280-all-demos  

Even locally, for example, the theme builder is taking very long to load on this branch:

```py
import gradio as gr

gr.themes.builder()
```

**Comment by abidlabs** [comment]
Just saw your comment -- I had a hunch it was to do with the `.load()` event as well, because demos without the `.load()` event (like `kitchen_sink`) seem to load very quickly.

**Comment by abidlabs** [comment]
Tested a bunch of demos locally -- leaving performance issues aside, everything looks good (events are triggered correct, as are chained events, and errors).

**Comment by pngwn** [comment]
Hmm, I'm getting terrible performance everywhere. Looking into it.

**Review by aliabid94 — COMMENTED** [review]
Two issues I've had so far:
- Examples don't seem to work, try sentence_buiolder demo
- Video play/pause don't seem to trigger in the event_trigger demo.

File: js/annotatedimage/static/AnnotatedImage.svelte [hunk 1]
```diff
@@ -36,14 +41,15 @@
 		if (value !== old_value) {
 			old_value = value;
 			dispatch("change");
```
[the remark this answers is not in the document]
Comment (pngwn) [reply js/annotatedimage/static/AnnotatedImage.svelte:43]: In one or two cases we are (when we are using component events for non-gradio events, local stuff) but I'll take another pass through and double check everything.

**Comment by pngwn** [comment]
Thanks @aliabid94! I'll take a look at those. I think i forgot to change examples altogether.

**Comment by pngwn** [comment]
Okay @abidlabs @aliabid94. I think I've address the main issue, this is ready to be tested again.

I'm just going to clean everything up, revert the demos, and get the tests passing again.

File: js/annotatedimage/static/AnnotatedImage.svelte (hunk 1 above)
[the remark this answers is not in the document]
Comment (pngwn) [reply js/annotatedimage/static/AnnotatedImage.svelte]: I think I've removed all of these. Let me know if i missed any,

**Comment by abidlabs** [comment]
Thanks for the helpful description @pngwn! The preview Space definitely loads faster. Noticed a few small issues:

1. The babylon logo is appearing in the top left of the preview Space

image


2. The theme builder isn't loading for me. If you run:

```py
import gradio as gr

gr.themes.builder()
```

it stays "loading" forever. Nothing in the console logs. 

-- Will keep testing --

**Comment by pngwn** [comment]
Yea I saw the Babylon loader. Not sure what I did there but will fix. I didn't test the theme builder, will take a look at that.

**Comment by abidlabs** [comment]
3. Gap between dropdown box and choices. Run the `native_plots` demo:

image

vs on `main`:

image

**Comment by pngwn** [comment]
I have tweaked the test a little bit to get things working again.

The units tests helpers are now a bit nicer. The `render` function now returns two additional methods:

- `listen` which takes an event name (this is typed) and returns a mock you can asserts against.
- `wait_for_event` which is identical to `listen` except you can `await` and it won't resolve until the event has been fired:

```ts
const { listen, wait_for_event } = await render(Component, props);

const input_mock = listen('input')
// enter some input
expect(input_mock).toHaveBeenCalled();

event.upload(item, file);
const change_mock = await wait_for_event("change");
expect(change_mock).toHaveBeenCalled();
```

**Comment by pngwn** [comment]
I'm hoping everything passes now, I just need to fix that babylon issue. 

@abidlabs, I can't reproduce the theme builder thing, could you try it again? Or if someone else could try it too.

**Comment by dawoodkhan82** [comment]
Pulling `MarkdownCode` into the markdown component is much cleaner. thanks for these changes!

**Comment by abidlabs** [comment]
Btw just pushed up a fix for some merge conflicts & the demos notebook test

**Comment by abidlabs** [comment]
>  Pulling MarkdownCode into the markdown component is much cleaner. thanks for these changes!

Agree with @dawoodkhan82 this is awesome! Not only is it cleaner, but we get support for syntax highlighting in code snippets in `gr.Markdown()` and `gr.DataFrame` which we didn't have before. 

Couple of nits regarding that:

1. The copy buttons in code snippets in `gr.Dataframe` don't work (aren't clickable) and if the language is typed, there's a weird beige background:

image

There's also a font difference between the code if you pass in a language versus if you don't specify the language. 

Code:

````py
import gradio as gr

code = """
```
def test(x):
  return x
```
"""

code_python = """
```py
def test(x):
  return x
```
"""

with gr.Blocks() as demo:
    gr.DataFrame([["regular", code], ["python", code_python]], datatype="markdown")
    
demo.launch()
````

2. The Markdown component looks much better with code snippets, although the font difference is still there:

image


````py
import gradio as gr

code = """
# abc
Here's python code:
```py
def test(x):
  return x
```
and here's regular code:
```
def test(x):
  return x
```

"""

with gr.Blocks() as demo:
    gr.Markdown(code)
    
demo.launch()
````

3. We used to have this font-difference problem with `gr.Chatbot` but then we pushed a patch a while back. The problem seems to have returned.

image


````py
import gradio as gr


with gr.Blocks() as demo:
    gr.Chatbot([(code, code_python)])
    
demo.launch()
````

Personally I prefer the font of the code snippet when the language is not specified, but whichever we choose, let's make it consistent.

**Comment by abidlabs** [comment]
Btw I'm still seeing the babylon logo in the preview Space, but the theme builder issue is not occurring for me anymore, nor is the dropdown issue. Besides that, lgtm @pngwn

**Comment by pngwn** [comment]
Thanks for the reviews folks! 

I think the font issues are due to the prism css, which only applies when there is a language. I'll iron out those inconsistencies, same for the strange background. 

I think the copy button not working for the Dataframe is just because I forgot to add some code, that's also an easy fix.

**Comment by abidlabs** [comment]
One bug I noticed is that the text in a chatbot is now being separated into separate lines:

image

image

**Comment by pngwn** [comment]
oh

**Review by abidlabs — APPROVED** [review]
LGTM thank you @pngwn for addressing the feedback and great PR!

**Pull request merged**